hypervisor/dm/vpci/vpci.c

```diff
@@ -370,30 +370,89 @@ static void vpci_deinit_pt_dev(struct pci_vdev *vdev)
 	deinit_vmsi(vdev);
 }
 
-static int32_t vpci_write_pt_dev_cfg(struct pci_vdev *vdev, uint32_t offset,
-		uint32_t bytes, uint32_t val)
+struct cfg_header_perm {
+	uint16_t pt_mask;	/* Whether per bit for per 4B register is pass through or virtualize */
+	uint16_t ro_mask;	/* Whether per bit for per 4B register is read-only or writable */
+};
+
+static const struct cfg_header_perm cfg_hdr_perm = {
+	/* Only Command (0x04-0x05) and Status (0x06-0x07) Registers are pass through */
+	.pt_mask = 0x0002U,
+	/* Command (0x04-0x05) and Status (0x06-0x07) Registers and
+	 * Base Address Registers (0x10-0x27) are writable */
+	.ro_mask = (uint16_t)~0x03f2U
+};
+
+
+/*
+ * @pre offset + bytes < PCI_STD_HEADER_LENGTH
+ */
+static void read_cfg_header_cfg(const struct pci_vdev *vdev,
+		uint32_t offset, uint32_t bytes, uint32_t *val)
+{
+	if (vbar_access(vdev, offset)) {
+		/* bar access must be 4 bytes and offset must also be 4 bytes aligned */
+		if ((bytes == 4U) && ((offset & 0x3U) == 0U)) {
+			*val = pci_vdev_read_bar(vdev, pci_bar_index(offset));
+		} else {
+			*val = ~0U;
+		}
+	} else {
+		if (bitmap32_test(offset >> 2U, (const volatile uint32_t *)&cfg_hdr_perm.pt_mask)) {
```

Comment:
do we need this type casting?

Reply:
Yes if we define this mask as uint16_t. I will change it to uint32_t although we only use the low 16 bits for MISRA-C.